File: frontend/lib/schemas/patient_ng_staff.ts
/**
 * NG スタッフ (患者 × スタッフ割当禁止) zod schemas.
 *
 * 設計書 `docs/plans/patient-ng-staff-design.md` §4 の API 契約に 1:1 対応。
 * Backend `backend/app/api/v1/patients` の ng-staff エンドポイントと一致させる。
 *
 * 概念 (§2):
 *   - 患者ごとに「このスタッフは割り当てない」恒久ルールを持つ (複数名可・理由メモ任意)。
 *   - 強度は性別制限 (`Patient.sex_restriction`) と同格。エンジンはハード制約、
 *     人手操作はブロックせず確認ダイアログで通す (§7)。
 */
import { z } from 'zod';

// ---------------------------------------------------------------------------
// GET /api/v1/patients/{patient_id}/ng-staff — response item (§4-1)
// ---------------------------------------------------------------------------

export const patientNgStaffReadSchema = z.object({
  staff_id: z.string().uuid(),
  /** BE は解決済みの氏名を載せる。欠落・null は「(氏名未登録)」表示に落とす。 */
  staff_name: z.string().nullable().optional(),
  /** 退職 (soft delete) 済みスタッフでも行は残る → 表示側で「(退職)」注記。 */
  staff_is_deleted: z.boolean().default(false),
  note: z.string().nullable().optional(),
  decided_by_user_id: z.string().uuid().nullable().optional(),
  created_at: z.string().nullable().optional(),
});
export type PatientNgStaffRead = z.infer<typeof patientNgStaffReadSchema>;

export const patientNgStaffListSchema = z.array(patientNgStaffReadSchema);

/**
 * 閲覧系 (§8-2) の 1 行表示用ラベル。0 件は「なし」。
 * 退職済みスタッフは「(退職)」を添える。
 */
export function formatNgStaffNames(rows: readonly PatientNgStaffRead[]): string {
  if (rows.length === 0) return 'なし';
  return rows
    .map((r) => `${r.staff_name ?? '(氏名未登録)'}${r.staff_is_deleted ? '(退職)' : ''}`)
    .join('・');
}

// ---------------------------------------------------------------------------
// PUT /api/v1/patients/{patient_id}/ng-staff/{staff_id} — request (§4-1)
// ---------------------------------------------------------------------------

export const patientNgStaffUpsertSchema = z.object({
  note: z.string().nullable(),
});
export type PatientNgStaffUpsert = z.infer<typeof patientNgStaffUpsertSchema>;

// ---------------------------------------------------------------------------
// §7-2: 手動経路 (PATCH /courses/{id}) の 422 acknowledge プロトコル
//   { detail: { code: 'constraint_confirmation_required', warnings: [...] } }
// ---------------------------------------------------------------------------

export const CONSTRAINT_WARNING_KINDS = ['gender', 'ng_staff'] as const;
export type ConstraintWarningKind = (typeof CONSTRAINT_WARNING_KINDS)[number];

export const constraintWarningSchema = z.object({
  kind: z.enum(CONSTRAINT_WARNING_KINDS),
  patient_id: z.string(),
  patient_name: z.string().nullable().optional(),
  staff_id: z.string(),
  staff_name: z.string().nullable().optional(),
  note: z.string().nullable().optional(),
});
export type ConstraintWarning = z.infer<typeof constraintWarningSchema>;

export const CONSTRAINT_CONFIRMATION_CODE = 'constraint_confirmation_required';

export const constraintConfirmationDetailSchema = z.object({
  code: z.literal(CONSTRAINT_CONFIRMATION_CODE),
  warnings: z.array(constraintWarningSchema),
});
export type ConstraintConfirmationDetail = z.infer<typeof constraintConfirmationDetailSchema>;

/**
 * ApiError.body (422) から確認要求の詳細を安全に取り出す。BE は
 * `{ detail: { code: 'constraint_confirmation_required', warnings: [...] } }` で返す。
 * 形が違えば null (= 通常のエラー表示にフォールバック)。
 *
 * 手本: `parseOverlapDetail` (`lib/schemas/trainee_accompaniment.ts`)。
 */
export function parseConstraintConfirmationDetail(
  body: unknown,
): ConstraintConfirmationDetail | null {
  if (typeof body !== 'object' || body === null) return null;
  const detail = (body as { detail?: unknown }).detail;
  const parsed = constraintConfirmationDetailSchema.safeParse(detail);
  return parsed.success ? parsed.data : null;
}

File: backend/app/schemas/patient_ng_staff.py
"""Pydantic schemas for patient_ng_staff (NG スタッフ).

正典設計書: ``docs/plans/patient-ng-staff-design.md`` §4-1.

* 行が存在する = NG (「NG のみ行を作る」方式). ``patient_same_address_link`` の
  schema 流儀を踏襲する.
* ``staff_name`` は ORM に無い派生項目 (staff テーブルを JOIN して endpoint 側で
  組み立てる) のため ``from_attributes`` ではなく明示構築する.
* ``staff_is_deleted`` は「退職済みスタッフの既存 NG 行」を表示側で「(退職)」と
  注記するためのフラグ. 退職しても行は残す (設計書 §4-1 バリデーション).
"""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class PatientNgStaffRead(BaseModel):
    """GET レスポンス 1 件 (= NG スタッフ 1 名)."""

    model_config = ConfigDict(extra="forbid")

    staff_id: UUID
    staff_name: str
    # 退職 (staff.deleted_at IS NOT NULL) 済みでも行は残す. 表示側で「(退職)」注記.
    staff_is_deleted: bool = False
    note: str | None = None
    decided_by_user_id: UUID | None = None
    created_at: datetime


class StaffNgPatientRead(BaseModel):
    """逆引き (``GET /staff/{staff_id}/ng-patients``) レスポンス 1 件.

    「このスタッフを NG 指定している患者」1 名分. スタッフ詳細の閲覧サマリ
    (設計書 §8-2 Phase 2「スタッフ詳細の逆引き」) 専用の派生ビュー.
    ``patient_name`` は patients を JOIN して都度解決する (二重化しない).
    削除済み患者 (``patients.deleted_at IS NOT NULL``) は含めない.
    """

    model_config = ConfigDict(extra="forbid")

    patient_id: UUID
    patient_name: str
    note: str | None = None
    created_at: datetime


class PatientNgStaffUpsert(BaseModel):
    """PUT リクエスト body (理由メモのみ).

    patient_id / staff_id は path param で受けるため body には含めない.
    ``note`` の長さ制限は同住所リンク (``PatientSameAddressLinkBase.note``) に
    合わせて設けない (DB 側も Text).
    """

    model_config = ConfigDict(extra="forbid")

    note: str | None = None
